File: expense-tracker/backend/storage/categories.js
const fs = require('fs').promises;
const path = require('path');
const { v4: uuidv4 } = require('uuid');

const DATA_DIR = process.env.DATA_PATH || path.join(__dirname, '..', 'data');
const CATEGORIES_FILE = path.join(DATA_DIR, 'categories.json');

// Default categories with colors
const DEFAULT_CATEGORIES = [
  { id: 'cat-groceries', name: 'Groceries', color: '#22c55e' },
  { id: 'cat-transport', name: 'Transport', color: '#3b82f6' },
  { id: 'cat-utilities', name: 'Utilities', color: '#f59e0b' },
  { id: 'cat-entertainment', name: 'Entertainment', color: '#8b5cf6' },
  { id: 'cat-dining', name: 'Dining', color: '#ef4444' },
  { id: 'cat-healthcare', name: 'Healthcare', color: '#06b6d4' },
  { id: 'cat-shopping', name: 'Shopping', color: '#ec4899' },
  { id: 'cat-other', name: 'Other', color: '#6b7280' },
];

// Ensure data directory exists
async function ensureDataDir() {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
  } catch (err) {
    // Directory already exists or can't be created
  }
}

// Initialize categories file with defaults if it doesn't exist
async function initCategories() {
  await ensureDataDir();
  try {
    await fs.access(CATEGORIES_FILE);
    // File exists, do nothing
  } catch (err) {
    // File doesn't exist, create with defaults
    await fs.writeFile(
      CATEGORIES_FILE,
      JSON.stringify({ categories: DEFAULT_CATEGORIES }, null, 2),
      'utf-8'
    );
  }
}

// Read all categories
async function getAllCategories() {
  await initCategories();
  const content = await fs.readFile(CATEGORIES_FILE, 'utf-8');
  const data = JSON.parse(content);
  return data.categories || [];
}

// Get a single category by ID
async function getCategoryById(categoryId) {
  const categories = await getAllCategories();
  return categories.find(c => c.id === categoryId) || null;
}

// Create a new category
async function createCategory({ name, color }) {
  const categories = await getAllCategories();
  
  // Check for duplicate name
  if (categories.some(c => c.name.toLowerCase() === name.toLowerCase())) {
    throw new Error('Category with this name already exists');
  }
  
  const newCategory = {
    id: `cat-${uuidv4().split('-')[0]}`,
    name: name.trim(),
    color: validateColor(color),
  };
  
  categories.push(newCategory);
  await fs.writeFile(
    CATEGORIES_FILE,
    JSON.stringify({ categories }, null, 2),
    'utf-8'
  );
  
  return newCategory;
}

// Update a category
async function updateCategory(categoryId, { name, color }) {
  const categories = await getAllCategories();
  const index = categories.findIndex(c => c.id === categoryId);
  
  if (index === -1) {
    throw new Error('Category not found');
  }
  
  // Check for duplicate name (excluding current category)
  if (name && categories.some(c => c.id !== categoryId && c.name.toLowerCase() === name.toLowerCase())) {
    throw new Error('Category with this name already exists');
  }
  
  if (name) categories[index].name = name.trim();
  if (color) categories[index].color = validateColor(color);
  
  await fs.writeFile(
    CATEGORIES_FILE,
    JSON.stringify({ categories }, null, 2),
    'utf-8'
  );
  
  return categories[index];
}

// Delete a category
async function deleteCategory(categoryId) {
  const categories = await getAllCategories();
  const index = categories.findIndex(c => c.id === categoryId);
  
  if (index === -1) {
    throw new Error('Category not found');
  }
  
  categories.splice(index, 1);
  await fs.writeFile(
    CATEGORIES_FILE,
    JSON.stringify({ categories }, null, 2),
    'utf-8'
  );
  
  return { deleted: categoryId };
}

// Validate hex color
function validateColor(color) {
  if (!color || typeof color !== 'string') {
    return '#6b7280'; // Default gray
  }
  // Allow #RGB or #RRGGBB format
  const hexRegex = /^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$/;
  if (hexRegex.test(color)) {
    // Normalize 3-digit hex to 6-digit
    if (color.length === 4) {
      const r = color[1];
      const g = color[2];
      const b = color[3];
      return `#${r}${r}${g}${g}${b}${b}`;
    }
    return color.toLowerCase();
  }
  return '#6b7280'; // Default gray for invalid colors
}

// Get expense count for a category across all years
async function getCategoryExpenseCount(categoryId, getAllYearsData) {
  const years = await getAllYearsData();
  let count = 0;
  
  for (const year of years) {
    for (const month of year.months) {
      count += month.expenses.filter(e => e.categoryId === categoryId).length;
    }
  }
  
  return count;
}

module.exports = {
  getAllCategories,
  getCategoryById,
  createCategory,
  updateCategory,
  deleteCategory,
  validateColor,
  getCategoryExpenseCount,
};

File: expense-tracker/backend-rust/src/storage/categories.rs
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use tokio::fs;
use uuid::Uuid;
use crate::models::Category;
use crate::error::AppError;

const DEFAULT_CATEGORIES: &[(&str, &str, &str)] = &[
    ("cat-groceries",    "Groceries",     "#22c55e"),
    ("cat-transport",    "Transport",     "#3b82f6"),
    ("cat-utilities",    "Utilities",     "#f59e0b"),
    ("cat-entertainment","Entertainment", "#8b5cf6"),
    ("cat-dining",       "Dining",        "#ef4444"),
    ("cat-healthcare",   "Healthcare",    "#06b6d4"),
    ("cat-shopping",     "Shopping",      "#ec4899"),
    ("cat-other",        "Other",         "#6b7280"),
];

#[derive(Clone)]
pub struct CategoriesStore {
    file_path: PathBuf,
    cache: Arc<Mutex<Vec<Category>>>,
}

impl CategoriesStore {
    pub async fn new(data_dir: &str) -> Result<Self, anyhow::Error> {
        let file_path = PathBuf::from(data_dir).join("categories.json");
        fs::create_dir_all(data_dir).await?;

        let categories = if file_path.exists() {
            let content = fs::read_to_string(&file_path).await?;
            let v: serde_json::Value = serde_json::from_str(&content)?;
            serde_json::from_value(v["categories"].clone())?
        } else {
            let defaults: Vec<Category> = DEFAULT_CATEGORIES.iter()
                .map(|(id, name, color)| Category {
                    id: id.to_string(),
                    name: name.to_string(),
                    color: color.to_string(),
                })
                .collect();
            let content = serde_json::to_string_pretty(&serde_json::json!({ "categories": defaults }))?;
            fs::write(&file_path, content).await?;
            defaults
        };

        Ok(Self { file_path, cache: Arc::new(Mutex::new(categories)) })
    }

    pub fn get_all(&self) -> Vec<Category> {
        self.cache.lock().unwrap().clone()
    }

    pub fn get_by_id(&self, id: &str) -> Option<Category> {
        self.cache.lock().unwrap().iter().find(|c| c.id == id).cloned()
    }

    pub async fn create(&self, name: String, color: String) -> Result<Category, AppError> {
        let cat = {
            let mut cache = self.cache.lock().unwrap();
            if cache.iter().any(|c| c.name.to_lowercase() == name.to_lowercase()) {
                return Err(AppError::Conflict("Category with this name already exists".to_string()));
            }
            let cat = Category {
                id: format!("cat-{}", &Uuid::new_v4().to_string()[..8]),
                name: name.trim().to_string(),
                color: Self::normalize_color(Some(&color)),
            };
            cache.push(cat.clone());
            cat
        }; // MutexGuard dropped here before await
        self.persist().await?;
        Ok(cat)
    }

    pub async fn update(&self, id: &str, name: Option<String>, color: Option<String>) -> Result<Category, AppError> {
        let updated = {
            let mut cache = self.cache.lock().unwrap();
            let idx = cache.iter().position(|c| c.id == id)
                .ok_or_else(|| AppError::NotFound("Category not found".to_string()))?;
            if let Some(ref n) = name {
                if cache.iter().enumerate().any(|(i, c)| i != idx && c.name.to_lowercase() == n.to_lowercase()) {
                    return Err(AppError::Conflict("Category with this name already exists".to_string()));
                }
                cache[idx].name = n.trim().to_string();
            }
            if let Some(ref c) = color {
                cache[idx].color = Self::normalize_color(Some(c));
            }
            cache[idx].clone()
        }; // MutexGuard dropped here before await
        self.persist().await?;
        Ok(updated)
    }

    pub async fn delete(&self, id: &str) -> Result<(), AppError> {
        {
            let mut cache = self.cache.lock().unwrap();
            let idx = cache.iter().position(|c| c.id == id)
                .ok_or_else(|| AppError::NotFound("Category not found".to_string()))?;
            cache.remove(idx);
        } // MutexGuard dropped here before await
        self.persist().await?;
        Ok(())
    }

    async fn persist(&self) -> Result<(), AppError> {
        let cats = { self.cache.lock().unwrap().clone() }; // guard dropped before await
        let content = serde_json::to_string_pretty(&serde_json::json!({ "categories": cats }))
            .map_err(|e| AppError::Storage(e.to_string()))?;
        fs::write(&self.file_path, content).await
            .map_err(|e| AppError::Storage(e.to_string()))
    }

    /// Normalise hex color: accepts #RGB → #RRGGBB, rejects invalid → default grey.
    pub fn normalize_color(color: Option<&str>) -> String {
        let Some(c) = color else { return "#6b7280".to_string() };
        let re = regex_lite::Regex::new(r"^#([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$").unwrap();
        if !re.is_match(c) { return "#6b7280".to_string() }
        if c.len() == 4 {
            let r = &c[1..2]; let g = &c[2..3]; let b = &c[3..4];
            format!("#{}{}{}{}{}{}", r,r,g,g,b,b).to_lowercase()
        } else {
            c.to_lowercase()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    async fn make_store(dir: &TempDir) -> CategoriesStore {
        CategoriesStore::new(dir.path().to_str().unwrap()).await.unwrap()
    }

    #[tokio::test]
    async fn test_defaults_created_on_first_load() {
        let temp = TempDir::new().unwrap();
        let store = make_store(&temp).await;
        let cats = store.get_all();
        assert!(!cats.is_empty(), "defaults should be seeded");
        assert!(cats.iter().any(|c| c.name == "Groceries"));
    }

    #[tokio::test]
    async fn test_create_and_get_category() {
        let temp = TempDir::new().unwrap();
        let store = make_store(&temp).await;
        let cat = store.create("Food".to_string(), "#ff0000".to_string()).await.unwrap();
        assert_eq!(cat.name, "Food");
        assert_eq!(cat.color, "#ff0000");
        assert!(store.get_by_id(&cat.id).is_some());
    }

    #[tokio::test]
    async fn test_duplicate_name_rejected() {
        let temp = TempDir::new().unwrap();
        let store = make_store(&temp).await;
        store.create("Unique".to_string(), "#aaaaaa".to_string()).await.unwrap();
        let result = store.create("unique".to_string(), "#bbbbbb".to_string()).await;
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn test_update_category() {
        let temp = TempDir::new().unwrap();
        let store = make_store(&temp).await;
        let cat = store.create("Old".to_string(), "#111111".to_string()).await.unwrap();
        let updated = store.update(&cat.id, Some("New".to_string()), Some("#222222".to_string())).await.unwrap();
        assert_eq!(updated.name, "New");
        assert_eq!(updated.color, "#222222");
    }

    #[tokio::test]
    async fn test_delete_category() {
        let temp = TempDir::new().unwrap();
        let store = make_store(&temp).await;
        let cat = store.create("Temp".to_string(), "#333333".to_string()).await.unwrap();
        store.delete(&cat.id).await.unwrap();
        assert!(store.get_by_id(&cat.id).is_none());
    }

    #[tokio::test]
    async fn test_color_normalisation() {
        assert_eq!(CategoriesStore::normalize_color(Some("#ABC")), "#aabbcc");
        assert_eq!(CategoriesStore::normalize_color(Some("#aabbcc")), "#aabbcc");
        assert_eq!(CategoriesStore::normalize_color(Some("notacolor")), "#6b7280");
        assert_eq!(CategoriesStore::normalize_color(None), "#6b7280");
    }
}
